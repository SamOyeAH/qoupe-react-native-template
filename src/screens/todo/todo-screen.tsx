import { FC, useLayoutEffect } from 'react';
import { Button, TextInput, View } from 'react-native';
import { useNavigation, useRoute } from '@react-navigation/native';
import { useLocalize } from '@hooks/localize';
import { FullScreenView } from '@components/full-screen-view';
import { useTodoScreen } from './todo-screen.hooks';
import styles from './todo-screen.styles';

export const TodoScreen: FC = () => {
  // TODO: find out how to type properly with @react-navigation
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const { params } = useRoute<{ params: { id: number } }>();
  const navigation = useNavigation();
  const { text, isDone, setText, saveTodo, changeTodoStatus, removeTodo } =
    useTodoScreen(params.id);
  const { t } = useLocalize();

  const handleSave = () => {
    saveTodo();
    navigation.goBack();
  };

  const handleToggle = () => {
    changeTodoStatus();
    navigation.goBack();
  };

  const handleDelete = () => {
    removeTodo();
    navigation.goBack();
  };

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => <Button onPress={handleSave} title={t('todo:save')} />,
    });
  }, [text]);

  return (
    <FullScreenView style={styles.root}>
      <TextInput
        testID="edit"
        style={styles.input}
        value={text}
        onChangeText={setText}
      />
      <View style={styles.buttons}>
        <View>
          <Button
            testID="delete"
            color="#CC0000"
            title={t('todo:delete')}
            onPress={handleDelete}
          />
        </View>
        <View>
          <Button
            testID="toggle"
            title={isDone ? t('todo:reopen') : t('todo:complete')}
            onPress={handleToggle}
          />
        </View>
      </View>
    </FullScreenView>
  );
};
