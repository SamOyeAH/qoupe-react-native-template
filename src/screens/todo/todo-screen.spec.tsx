import { render, fireEvent } from '@testing-library/react-native';
import { TodoScreen } from './todo-screen';

const mockGoBack = jest.fn();
const mockSetOptions = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack: () => mockGoBack(),
    setOptions: () => mockSetOptions(),
  }),
  useRoute: () => ({ params: { id: 1 } }),
}));

const mockUseTodoScreen = jest.fn();
jest.mock('./todo-screen.hooks', () => ({
  useTodoScreen: (id: number) => mockUseTodoScreen(id),
}));

describe('TodoScreen', () => {
  const mock = {
    text: 'text',
    isDone: false,
    setText: jest.fn(),
    saveTodo: jest.fn(),
    changeTodoStatus: jest.fn(),
    removeTodo: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Renders incomplete todo', () => {
    mockUseTodoScreen.mockReturnValue(mock);
    const { toJSON } = render(<TodoScreen />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('Renders completed todo', () => {
    mockUseTodoScreen.mockReturnValue({
      ...mock,
      isDone: true,
    });
    const { toJSON } = render(<TodoScreen />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('Handles todo delete', () => {
    mockUseTodoScreen.mockReturnValue(mock);
    const { getByTestId } = render(<TodoScreen />);

    fireEvent.press(getByTestId('delete'));
    expect(mock.removeTodo).toHaveBeenCalled();
    expect(mockGoBack).toHaveBeenCalled();
  });

  it('Handles todo toggle', () => {
    mockUseTodoScreen.mockReturnValue(mock);
    const { getByTestId } = render(<TodoScreen />);

    fireEvent.press(getByTestId('toggle'));
    expect(mock.changeTodoStatus).toHaveBeenCalled();
    expect(mockGoBack).toHaveBeenCalled();
  });

  it('Handles text change', () => {
    mockUseTodoScreen.mockReturnValue(mock);
    const { getByTestId } = render(<TodoScreen />);

    fireEvent.changeText(getByTestId('edit'), 'new text');
    expect(mock.setText).toHaveBeenCalledWith('new text');
  });
});
