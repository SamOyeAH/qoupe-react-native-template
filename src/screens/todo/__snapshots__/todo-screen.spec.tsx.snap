// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoScreen Renders completed todo 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingHorizontal": 16,
        "paddingVertical": 24,
      },
    ]
  }
>
  <TextInput
    onChangeText={[MockFunction]}
    style={
      Object {
        "borderBottomColor": "#aaa",
        "borderBottomWidth": 1,
        "paddingBottom": 8,
      }
    }
    testID="edit"
    value="text"
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 16,
      }
    }
  >
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="delete"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#CC0000",
                },
              ]
            }
          >
            todo:delete
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="toggle"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            todo:reopen
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`TodoScreen Renders incomplete todo 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingHorizontal": 16,
        "paddingVertical": 24,
      },
    ]
  }
>
  <TextInput
    onChangeText={[MockFunction]}
    style={
      Object {
        "borderBottomColor": "#aaa",
        "borderBottomWidth": 1,
        "paddingBottom": 8,
      }
    }
    testID="edit"
    value="text"
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 16,
      }
    }
  >
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="delete"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#CC0000",
                },
              ]
            }
          >
            todo:delete
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="toggle"
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            todo:complete
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
